data-processing/pdf/process_pdf.py

```diff
@@ -0,0 +1,236 @@
+            else:
+                index -= len(p['tokens'])
+        return None
+
+
+    def get_text(self,spans):
+        text = []
+        for s in spans:
+            direct = s.get('dehyphenizedText')
+            if direct:
+                text += direct
+            else:
+                for i in range(s['left'], s['right']):
+                    page, token = self.find_token(i)
+                    text.append(token['text'])
+        return ' '.join(text)
+
+    def union(self, bbox1: BoundingBoxModel, bbox2: BoundingBoxModel):
+        if not bbox1:
+            return bbox2
+        x1 = min(bbox1.left,bbox2.left)
+        y1 = min(bbox1.top,bbox2.top)
+        x2 = max(bbox1.left+bbox1.width, bbox2.left+bbox2.width)
+        y2 = max(bbox1.top+bbox1.height, bbox2.top+bbox2.height)
+        return BoundingBoxModel(page = bbox1.page,
+                                left = x1, top = y1, width = x2-x1, height=y2-y1)
+
+    @staticmethod
+    def should_combine(bbox1: BoundingBoxModel, bbox2: BoundingBoxModel):
+        return bbox1.page == bbox2.page and (
+            abs(bbox1.top - bbox2.top) < 4 # Same y-coordinate
+        ) and (
+            abs(bbox2.left - bbox1.left - bbox1.width) < 15 # To the right
+        )
+
+    def get_bounding_boxes(self, spans):
+        bboxes = []
+        for s in spans:
+            bbox = None
+            for i in range(s['left'], s['right']):
+                page, token = self.find_token(i)
+                page_number = page['pageNumber']
+                token_bbox = BoundingBoxModel(
+                    page = page_number,
+                    left = token['x'],
+                    top = token['y'],
+                    width = token['width'],
+                    height=token['height']
+                )
+                if not bbox:
+                    bbox = token_bbox
+                elif self.should_combine(bbox, token_bbox):
+                     bbox = self.union(bbox, token_bbox)
+                else:
+                    bboxes.append(bbox)
+                    bbox = token_bbox
+            bboxes.append(bbox)
+        return bboxes
+
+
+    def find_cited_paper(self, bib_item_title):
+        if self.references is None:
+            resp = requests.get(f"https://api.semanticscholar.org/v1/paper/{self.pdf_hash}")
+            if resp.ok:
+                self.references = resp.json()['references']
+            else:
+                logging.warning('Got status %s for paper %s',resp.status_code, pdf_hash)
+                self.references = []
+
+        max_similarity = 0.0
+        best_matching_paper = None
+        for reference_data in self.references:
+            reference_title = reference_data['title']
+            similarity = ngram_sim(reference_title, bib_item_title)
+            if similarity > 0.5 and similarity > max_similarity:
+                max_similarity = similarity
+                best_matching_paper = reference_data
```

Comment:
This paper looks pretty similar to what we have in `resolve_bibitems.py`. Do you think it makes sense to pull it out into a new method?

Reply:
I think that's probably not worth the effort, given that the logic is simple, and the two loops are iterating over different types of objects.